include/boost/geometry/algorithms/detail/tupled_output.hpp

```diff
@@ -527,6 +502,234 @@ struct output_geometry_access<GeometryOut, Tag, DefaultTag, DefaultTag>
+            tupled_output_tag,
+            typename geometry::tag<GeometryOut>::type
+        >
+{};
+
+
+template <typename Geometry1, typename Geometry2, typename TupledOut, bool IsFound, typename Tag>
+struct expect_output_assert_base;
+
+template <typename Geometry1, typename Geometry2, typename TupledOut, bool IsFound>
+struct expect_output_assert_base<Geometry1, Geometry2, TupledOut, IsFound, pointlike_tag>
+{
+    BOOST_MPL_ASSERT_MSG
+        (
+            IsFound, POINTLIKE_GEOMETRY_EXPECTED_IN_TUPLED_OUTPUT,
+            (types<Geometry1, Geometry2, TupledOut>)
+        );
+};
+
+template <typename Geometry1, typename Geometry2, typename TupledOut, bool IsFound>
+struct expect_output_assert_base<Geometry1, Geometry2, TupledOut, IsFound, linear_tag>
+{
+    BOOST_MPL_ASSERT_MSG
+    (
+        IsFound, LINEAR_GEOMETRY_EXPECTED_IN_TUPLED_OUTPUT,
+        (types<Geometry1, Geometry2, TupledOut>)
+    );
+};
+
+template <typename Geometry1, typename Geometry2, typename TupledOut, bool IsFound>
+struct expect_output_assert_base<Geometry1, Geometry2, TupledOut, IsFound, areal_tag>
+{
+    BOOST_MPL_ASSERT_MSG
+    (
+        IsFound, AREAL_GEOMETRY_EXPECTED_IN_TUPLED_OUTPUT,
+        (types<Geometry1, Geometry2, TupledOut>)
+    );
+};
+
+
+template <typename Geometry1, typename Geometry2, typename TupledOut, typename Tag>
+struct expect_output_assert
+    : expect_output_assert_base
+        <
+            Geometry1, Geometry2, TupledOut,
+            geometry::tuples::exists_if
+                <
+                    TupledOut,
+                    is_tag_same_as_pred<Tag>::template pred
+                >::value,
+            typename geometry::tag_cast
+                <
+                    Tag, pointlike_tag, linear_tag, areal_tag
+                >::type
+        >
+{};
+
+template <typename Geometry1, typename Geometry2, typename TupledOut>
+struct expect_output_assert<Geometry1, Geometry2, TupledOut, void>
+{};
+
+template
+<
+    typename Geometry1, typename Geometry2, typename TupledOut,
+    typename Tag1,
+    typename Tag2 = void,
+    typename Tag3 = void
+>
+struct expect_output
+    : expect_output_assert<Geometry1, Geometry2, TupledOut, Tag1>
+    , expect_output_assert<Geometry1, Geometry2, TupledOut, Tag2>
+    , expect_output_assert<Geometry1, Geometry2, TupledOut, Tag3>
+{};
+
+template
+<
+    typename Geometry1, typename Geometry2, typename TupledOut,
+    typename Tag1, typename Tag2
+>
+struct expect_output<Geometry1, Geometry2, TupledOut, Tag1, Tag2, void>
+    : expect_output_assert<Geometry1, Geometry2, TupledOut, Tag1>
+    , expect_output_assert<Geometry1, Geometry2, TupledOut, Tag2>
+{};
+
+template
+<
+    typename Geometry1, typename Geometry2, typename TupledOut,
+    typename Tag1
+>
+struct expect_output<Geometry1, Geometry2, TupledOut, Tag1, void, void>
+    : expect_output_assert<Geometry1, Geometry2, TupledOut, Tag1>
+{};
+
+
+template <typename CastedTag>
+struct casted_tag_to_single_tag;
```

Comment:
Names of casts are typically suffixed with `cast`. Can you maybe rename this to `to_single_tag_cast` or `single_tag_cast` ?

Reply:
This is not a cast. It's a metafunction translating/mapping from `CastedTag` (dispatch template argument which can be pointlike_tag, linear_tag, areal_tag) to tag of a "single" geometry which is an element of multi-geometry (point_tag, linestring_tag, polygon_tag respectively).

Reply:
I renamed it to `single_tag_from_base_tag` to avoid "cast" word. Is that name better?